private/rules/kt_jvm_export.bzl

```diff
@@ -0,0 +1,84 @@
+load("@io_bazel_rules_kotlin//kotlin:jvm.bzl", "kt_jvm_library")
```

Comment:
Agh, the test failures in the examples CI pipeline is due to this load line adding `@io_bazel_rules_kotlin` as an implicit dependency of everyone using rules_jvm_external by loading `//:defs.bzl` -> `kt_jvm_export`, regardless of whether they're using `kt_jvm_export` or not. This is due to the non-transitive nature of repository rule deps.

This means we'll need to add rules_kotlin to //:repository.bzl to put `io_bazel_rules_kotlin` into the namespace and force all RJE users to call the setup/repo statements in their WORKSPACE (https://github.com/bazelbuild/rules_jvm_external#usage), instead of a single `http_archive` call. This is a backwards incompatible change.

Another way is to store kt_jvm_export under a different top level "defs.bzl" file, maybe "//:defs_with_deps.bzl", so users of those special symbols have to load the deps/setup macro (https://github.com/bazelbuild/rules_jvm_external/blob/master/repositories.bzl). This is a backwards compatible change.

cc @shs96c wdyt? I prefer the second approach.

Reply:
We already broke the "just use a single `http_archive`" intent when we added a dependency on `google-cloud-storage` to the exporter binary. That ship has already sailed. Adding `io_bazel_rules_kotlin` to `repositories.bzl` is the better choice in this case, since it fits with the workflow we've already made people follow, and which is documented in the readme.

Even if that weren't the case, asking users to be aware of which of our rules have third party dependencies and which don't seems like an unfair loading of responsibility, especially since a rule can start with no deps and then acquire some (just as `java_export`'s publish rule did)

Reply:
Sounds good. It was a soft requirement to add those lines, but this would cement that as a hard requirement. Let's just bite the bullet and do it then. 

@deannagarcia to summarize, there's a couple of things here:

1) Add rules_kotlin to repositories.bzl, and do any necessary setup for rules_kotlin in setup.bzl.
2) Update all the examples to load repositories.bzl and setup.bzl.

That should make CI happy.

Reply:
If the rules_kotlin setup gets complicated, please feel free to ask for a hand, either here or on the Slack channel.

Reply:
Okay I can get started on this, thanks for all the help! What's the best way for me to test my implementation to see if it will pass the CI test?

Reply:
`buildkite/rules-jvm-external-examples` is the failing CI job that builds and tests every project in `examples`, so your implementation should result in some changes to each example's workspace to add the repo and setup bzl to make it pass.

Reply:
I tried to do the steps and it is working on my local machine but all the buildkite tests are still failing. Can someone take a look and see if they have ideas on how to fix the tests?

Reply:
Have you tried to repro the CI-issued bazel commands locally? 

I can reproduce the error from `examples/simple`:

```
jingwen-macbookpro% cd examples/simple
jingwen-macbookpro% bazelisk test ...
2022/01/21 11:12:25 Downloading https://releases.bazel.build/5.0.0/release/bazel-5.0.0-darwin-x86_64...
Extracting Bazel installation...
Starting local Bazel server and connecting to it...
ERROR: Failed to load Starlark extension '@io_bazel_rules_kotlin//kotlin:jvm.bzl'.
Cycle in the workspace file detected. This indicates that a repository is used prior to being defined.
The following chain of repository dependencies lead to the missing definition.
 - @io_bazel_rules_kotlin
This could either mean you have to add the '@io_bazel_rules_kotlin' repository with a statement like `http_archive` in your WORKSPACE file (note that transitive dependencies are not added automatically), or move an existing definition earlier in your WORKSPACE file.
ERROR: cycles detected during target parsing
INFO: Elapsed time: 2.459s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded)
FAILED: Build did NOT complete successfully (0 packages loaded)
```

The issue is with this line:

https://github.com/bazelbuild/rules_jvm_external/blob/56246b52aceb428b7a065237d6f6c332070536e2/repositories.bzl#L1

Since `//:defs.bzl` now loads from rules_kotlin, the load statement is called before rules_kotlin is defined. I tried to rearrange the repo declarations, and I think we now need 3 steps/macros:

1. http_archive for rules_kotlin
2. call `@rules_jvm_external//:defs.bzl#maven_install` to create rules_jvm_external_deps
3. call `@rules_jvm_external_deps//:defs.bzl#pinned_maven_install` from rules_jvm_external_deps

These 3 steps have to be sequential, and cannot be in the same .bzl file. Alternatively, if we go down the route of declaring kt_jvm_export in a separate top level (e.g.) `kt_defs.bzl` file, then this wouldn't be a problem, I think.

There are a few more errors.

From: https://buildkite.com/bazel/rules-jvm-external/builds/2479#b813ab50-2257-404e-a022-e3f1e2c2168d

```
(19:31:17) ERROR: /workdir/tests/integration/kt_jvm_export/BUILD:3:1: in kt_jvm_library rule //tests/integration/kt_jvm_export:test-lib:
--
  | Traceback (most recent call last):
  | File "/workdir/tests/integration/kt_jvm_export/BUILD", line 3
  | kt_jvm_library(name = 'test-lib')
  | File "/var/lib/buildkite-agent/.cache/bazel/_bazel_buildkite-agent/ec321eb2cc2d0f8f91b676b6d4c66c29/external/io_bazel_rules_kotlin/kotlin/internal/jvm/impl.bzl", line 233, in kt_jvm_library_impl
  | _make_providers(ctx, _kt_jvm_produce_jar_actions(c...))
  | File "/var/lib/buildkite-agent/.cache/bazel/_bazel_buildkite-agent/ec321eb2cc2d0f8f91b676b6d4c66c29/external/io_bazel_rules_kotlin/kotlin/internal/jvm/impl.bzl", line 235, in _make_providers
  | _kt_jvm_produce_jar_actions(ctx, "kt_jvm_library")
  | File "/var/lib/buildkite-agent/.cache/bazel/_bazel_buildkite-agent/ec321eb2cc2d0f8f91b676b6d4c66c29/external/io_bazel_rules_kotlin/kotlin/internal/jvm/compile.bzl", line 532, in _kt_jvm_produce_jar_actions
  | java_common.pack_sources(ctx.actions, output_source_jar = c..., <3 more arguments>)
  | parameter 'output_jar' has no default value, for call to method pack_sources(actions, output_jar, sources = [], source_jars = [], java_toolchain, host_javabase) of 'java_common'
  | (19:31:18) ERROR: Analysis of target '//tests/integration/kt_jvm_export:check-deploy-env' failed; build aborted: Analysis of target '//tests/integration/kt_jvm_export:test-lib' failed; build aborted
```

I'm also getting a different error locally with `bazel test //...`

```
ERROR: /Users/jingwen/code/rules_jvm_external/BUILD:9:8: Generating proto for Starlark doc for defs failed: (Exit 1): stardoc failed: error executing command
  (cd /private/var/tmp/_bazel_jingwen/5e64b778ca5f4ff8a6ca994b3f33d8f8/sandbox/darwin-sandbox/566/execroot/rules_jvm_external && \
  exec env - \
  bazel-out/host/bin/external/io_bazel_stardoc/stardoc/stardoc '--input=//:defs.bzl' '--workspace_name=' '--symbols=javadoc' '--symbols=java_export' '--symbols=maven_install' '--dep_roots=.' '--dep_roots=external/' '--output=bazel-out/darwin-fastbuild/bin/defs.raw')
Execution platform: @local_config_platform//:host

Use --sandbox_debug to see verbose messages from the sandbox
Stardoc documentation generation failed: File defs.bzl imported '//private/rules:kt_jvm_export.bzl', yet private/rules/kt_jvm_export.bzl was not found, even at roots [., external/].
```